File: Data Structures/Python/Circular Doubly.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None

class CircularDoublyLinkedList:
    def __init__(self):
        self.head = None

    def insert_beginning(self, value):
        new_node = Node(value)
        if not self.head:
            new_node.next = new_node  # Circular reference to itself
            new_node.prev = new_node
        else:
            temp = self.head
            while temp.next != self.head:
                temp = temp.next
            temp.next = new_node
            new_node.prev = temp
        new_node.next = self.head
        self.head.prev = new_node
        self.head = new_node
        print("Element inserted at the beginning successfully.")

    def delete_beginning(self):
        if not self.head:
            print("List is empty. Cannot delete from the beginning.")
            return

        temp = self.head
        while temp.next != self.head:
            temp = temp.next

        if temp == self.head:
            self.head = None
        else:
            temp.next = self.head.next
            self.head.next.prev = temp
            self.head = temp.next

        print("Element deleted from the beginning successfully.")

    def insert_end(self, value):
        new_node = Node(value)

        if not self.head:
            new_node.next = new_node  # Circular reference to itself
            new_node.prev = new_node
            self.head = new_node
        else:
            temp = self.head
            while temp.next != self.head:
                temp = temp.next
            temp.next = new_node
            new_node.prev = temp

        new_node.next = self.head
        self.head.prev = new_node
        print("Element inserted at the end successfully.")

    def delete_end(self):
        if not self.head:
            print("List is empty. Cannot delete from the end.")
            return

        temp = self.head
        prev = None

        while temp.next != self.head:
            prev = temp
            temp = temp.next

        if temp == self.head:
            self.head = None
        else:
            prev.next = temp.next
            temp.next.prev = prev

        print("Element deleted from the end successfully.")

    def insert_at_position(self, value, position):
        if position < 1:
            print("Invalid position. Element not inserted.")
            return

        if position == 1:
            self.insert_beginning(value)
            return

        new_node = Node(value)

        temp = self.head
        for i in range(1, position - 1):
            if not temp:
                print("Invalid position. Element not inserted.")
                return
            temp = temp.next

        new_node.next = temp.next
        new_node.prev = temp
        if temp.next != self.head:
            temp.next.prev = new_node
        temp.next = new_node

        print(f"Element inserted at position {position} successfully.")

    def delete_at_position(self, position):
        if not self.head:
            print("List is empty. Cannot delete from the specified position.")
            return

        if position < 1:
            print("Invalid position. Element not deleted.")
            return

        temp = self.head
        prev = None

        for i in range(1, position):
            prev = temp
            temp = temp.next

        if temp == self.head:
            if temp.next == self.head:
                self.head = None
            else:
                prev.next = temp.next
                temp.next.prev = prev
                self.head = prev.next
        else:
            prev.next = temp.next
            temp.next.prev = prev

        print(f"Element deleted from position {position} successfully.")

    def display_list(self):
        if not self.head:
            print("List is empty.")
            return

        temp = self.head

        while True:
            print(temp.data, end=" ")
            temp = temp.next
            if temp == self.head:
                break

        print()

if __name__ == "__main__":
    circular_doubly_linked_list = CircularDoublyLinkedList()

    while True:
        print("\n***** Circular Doubly Linked List Operations *****")
        print("1. Insert Beginning")
        print("2. Delete Beginning")
        print("3. Insert End")
        print("4. Delete End")
        print("5. Insert at Position")
        print("6. Delete at Position")
        print("7. Display List")
        print("8. Exit")

        choice = int(input("Enter your choice: "))

        if choice == 1:
            value = int(input("Enter the value to insert: "))
            circular_doubly_linked_list.insert_beginning(value)
        elif choice == 2:
            circular_doubly_linked_list.delete_beginning()
        elif choice == 3:
            value = int(input("Enter the value to insert: "))
            circular_doubly_linked_list.insert_end(value)
        elif choice == 4:
            circular_doubly_linked_list.delete_end()
        elif choice == 5:
            value = int(input("Enter the value to insert: "))
            position = int(input("Enter the position to insert: "))
            circular_doubly_linked_list.insert_at_position(value, position)
        elif choice == 6:
            position = int(input("Enter the position to delete: "))
            circular_doubly_linked_list.delete_at_position(position)
        elif choice == 7:
            circular_doubly_linked_list.display_list()
        elif choice == 8:
            print("Exiting program. Goodbye!")
            break
        else:
            print("Invalid choice. Please enter a valid option.")

File: Data Structures/C/Circular Doubly.c
#include <stdio.h>
#include <stdlib.h>

struct Node{
    int data;
    struct Node* next;
    struct Node* prev;
};

struct Node* head = NULL;

void insertBeginning(int value){
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = value;
    newNode->next = head;
    newNode->prev = NULL;

    if (head == NULL) {
        newNode->next = newNode;  // Circular reference to itself
        newNode->prev = newNode;
    } else {
        struct Node* temp = head;
        while (temp->next != head) {
            temp = temp->next;
        }
        temp->next = newNode;
        newNode->prev = temp;
    }

    head = newNode;
    printf("Element inserted at the beginning successfully.\n");
}

void deleteBeginning() {
    if (head == NULL) {
        printf("List is empty. Cannot delete from the beginning.\n");
        return;
    }

    struct Node* temp = head;
    while (temp->next != head) {
        temp = temp->next;
    }

    if (temp == head) {
        free(head);
        head = NULL;
    } else {
        temp->next = head->next;
        head->next->prev = temp;
        free(head);
        head = temp->next;
    }

    printf("Element deleted from the beginning successfully.\n");
}

void insertEnd(int value) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = value;

    if (head == NULL) {
        newNode->next = newNode;  // Circular reference to itself
        newNode->prev = newNode;
        head = newNode;
    } else {
        struct Node* temp = head;
        while (temp->next != head) {
            temp = temp->next;
        }
        temp->next = newNode;
        newNode->prev = temp;
    }

    newNode->next = head;
    head->prev = newNode;
    printf("Element inserted at the end successfully.\n");
}

void deleteEnd() {
    if (head == NULL) {
        printf("List is empty. Cannot delete from the end.\n");
        return;
    }

    struct Node* temp = head;
    struct Node* prev = NULL;

    while (temp->next != head) {
        prev = temp;
        temp = temp->next;
    }

    if (temp == head) {
        free(head);
        head = NULL;
    } else {
        prev->next = temp->next;
        temp->next->prev = prev;
        free(temp);
    }

    printf("Element deleted from the end successfully.\n");
}

void insertAtPosition(int value, int position) {
    if (position < 1) {
        printf("Invalid position. Element not inserted.\n");
        return;
    }

    if (position == 1) {
        insertBeginning(value);
        return;
    }

    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = value;

    struct Node* temp = head;
    for (int i = 1; i < position - 1; i++) {
        if (temp == NULL) {
            printf("Invalid position. Element not inserted.\n");
            free(newNode);
            return;
        }
        temp = temp->next;
    }

    newNode->next = temp->next;
    newNode->prev = temp;
    if (temp->next != head) {
        temp->next->prev = newNode;
    }
    temp->next = newNode;

    printf("Element inserted at position %d successfully.\n", position);
}

void deleteAtPosition(int position) {
    if (head == NULL) {
        printf("List is empty. Cannot delete from the specified position.\n");
        return;
    }

    if (position < 1) {
        printf("Invalid position. Element not deleted.\n");
        return;
    }

    struct Node* temp = head;
    struct Node* prev = NULL;

    for (int i = 1; i < position; i++) {
        prev = temp;
        temp = temp->next;
    }

    if (temp == head) {
        if (temp->next == head) {
            free(head);
            head = NULL;
        } else {
            prev->next = temp->next;
            temp->next->prev = prev;
            free(temp);
            head = prev->next;
        }
    } else {
        prev->next = temp->next;
        temp->next->prev = prev;
        free(temp);
    }

    printf("Element deleted from position %d successfully.\n", position);
}

void displayList() {
    if (head == NULL) {
        printf("List is empty.\n");
        return;
    }

    struct Node* temp = head;

    do {
        printf("%d ", temp->data);
        temp = temp->next;
    } while (temp != head);

    printf("\n");
}

int main() {
    int choice, value, position;

    do {
        printf("\n***** Circular Doubly Linked List Operations *****\n");
        printf("1. Insert Beginning\n");
        printf("2. Delete Beginning\n");
        printf("3. Insert End\n");
        printf("4. Delete End\n");
        printf("5. Insert at Position\n");
        printf("6. Delete at Position\n");
        printf("7. Display List\n");
        printf("8. Exit\n");

        printf("Enter your choice: ");
        scanf("%d", &choice);

        switch (choice) {
            case 1:
                printf("Enter the value to insert: ");
                scanf("%d", &value);
                insertBeginning(value);
                break;
            case 2:
                deleteBeginning();
                break;
            case 3:
                printf("Enter the value to insert: ");
                scanf("%d", &value);
                insertEnd(value);
                break;
            case 4:
                deleteEnd();
                break;
            case 5:
                printf("Enter the value to insert: ");
                scanf("%d", &value);
                printf("Enter the position to insert: ");
                scanf("%d", &position);
                insertAtPosition(value, position);
                break;
            case 6:
                printf("Enter the position to delete: ");
                scanf("%d", &position);
                deleteAtPosition(position);
                break;
            case 7:
                displayList();
                break;
            case 8:
                printf("Exiting program. Goodbye!\n");
                break;
            default:
                printf("Invalid choice. Please enter a valid option.\n");
        }
    } while (choice != 8);

    return 0;
}
